src/inmanta/deploy/state.py

```diff
@@ -210,66 +215,193 @@ def are_dependencies_compliant(self, resource: ResourceIdStr) -> bool:
+            """
+            if resource in is_blocked:
+                # Is blocked for a know root
+                return False
 
-        :param resource: The resource that became defined.
-        :param details: The details of resource.
-        """
+            my_state = self.resource_state[resource]
+            if my_state.blocked is BlockedStatus.YES:
+                # The resource is already blocked.
+                return False
 
-        # Cache used by the `update_blocked_status()` method to prevent the expensive check on the blocked
-        # status of the requirements of a resource. If the dictionary contains key value pair (X, Y), it
-        # indicates that resource Y is a requirement of resource X that was blocked, last we checked
-        known_blockers_cache: dict[ResourceIdStr, ResourceIdStr] = {}
+            if my_state.status is ComplianceStatus.UNDEFINED:
+                assert False  # TODO: remove, should not happen
+
+            blocked_dependency: ResourceIdStr | None = None
+            # one blocked requires, evidence we are blocked
+
+            if resource in known_blockers_cache:
+                # First check the blocked status of the cached known blocker for improved performance.
+                known_blocker: ResourceIdStr = known_blockers_cache[resource]
+                if self.resource_state[known_blocker].blocked is BlockedStatus.YES:
+                    blocked_dependency = known_blocker
+                else:
+                    # Cache is out of date. Clear cache item.
+                    del known_blockers_cache[resource]
+
+            if blocked_dependency is None:
+                # Perform more expensive call by traversing all requirements of resource.
+                blocked_dependency = next(
+                    (r for r in self.requires.get(resource, set()) if self.resource_state[r].blocked is BlockedStatus.YES), None
+                )
+
+            if blocked_dependency is None:
+                return False
+
+            if blocked_dependency:
+                # Resource is blocked, because a dependency is blocked.
+                known_blockers_cache[resource] = blocked_dependency
+
+            if blocked_dependency in is_blocked:
+                # We know the root is undefined, hard block
+                is_blocked.add(resource)
+
+            # Unblock resource
+            self._block_resource_transitive(resource)
+            out_blocked.add(resource)
+            return True
+
+        for resource_id in verify_blocked:
+            # verify all potential blockers
+            blocked = update_blocked_status_to_blocked(resource_id)
+            # Keep track of this that have been blocked
+            if blocked:
+                propagate_blocked_work.append(resource_id)
+
+        # Propagate blockers once more
+        # We iterate differently, because the parents blocked status is relevant
+        propagate_blocked_work = list(verify_blocked)
```

Comment:
This doesn't make sense to me, why are we appending the blocked resources to `propagate_blocked_work` if we are just going to overwrite it?

Reply:
worklist algorthm! 

1. We consume form the list until it is empty
2. Here we initialize it with initial work
3. every iteration adds new work to the list (the code is defined above, but execute after this statement)


It is a way if doing (tail) recursion without ever running out of stack

Reply:
I don't get this part before we overwrite the `propagate_blocked_work` and start consuming from the list:
```
        for resource_id in verify_blocked:
            # verify all potential blockers
            blocked = update_blocked_status_to_blocked(resource_id)
            # Keep track of this that have been blocked
            if blocked: <------- Specifically this part and the line below
                propagate_blocked_work.append(resource_id)
```

Reply:
right!